### PR TITLE
Cosmos DB - Authorization policy

### LINKED ISSUE #324: CosmosDB E2E attachment tests fail with `not implemented` panic
Repro:

```bash
✔ 09:50 ~/src/rust/azure-sdk-for-rust/sdk/cosmos/src [master {origin/master}|✔] $ cargo test --features=test_e2e --test attachment_00
   Compiling azure_core v0.1.0 (/home/MINDFLAVOR/mindflavor/src/rust/azure-sdk-for-rust/sdk/core)
   Compiling azure_cosmos v0.1.0 (/home/MINDFLAVOR/mindflavor/src/rust/azure-sdk-for-rust/sdk/cosmos)
    Finished test [unoptimized + debuginfo] target(s) in 5.31s
     Running tests/attachment_00.rs (/home/MINDFLAVOR/mindflavor/src/rust/azure-sdk-for-rust/target/debug/deps/attachment_00-6688adeeb113b23c)

running 1 test
test attachment ... FAILED

failures:

---- attachment stdout ----
thread 'attachment' panicked at 'not implemented', /home/MINDFLAVOR/mindflavor/src/rust/azure-sdk-for-rust/sdk/core/src/headers/mod.rs:40:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    attachment

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 1.84s

error: test failed, to rerun pass '--test attachment_00'
```

This is due to https://github.com/Azure/azure-sdk-for-rust/blob/59de1214be443630d71dcd0f6c0275552df6fc3e/sdk/core/src/headers/mod.rs#L39-L41

Inserted by PR https://github.com/Azure/azure-sdk-for-rust/pull/204.

We should strive to correct it: we don't have a mean to test the crate otherwise.

### PR DESCRIPTION
**Pull request by MindFlavor**
This PR enhances the pipeline by adding the `AuthorizationPolicy` and migrates the preexisting CosmosDB pipeline operations to the new policy.

# AuthorizationPolicy

The `AuthorizationPolicy` cannot be in `Core` since every service uses a different authorization mechanism. So this PR proposes to have a specific policy in each module (they can potentially be shared in case of duplication). This PR covers Cosmos DB only. While CosmosDB does not support (yet) the bearer token authentication, it's possibile to authenticate specific *users* with shared key. 

The `AuthorizationPolicy` will handle the authentication by injecting the required headers to the request passing through the pipeline. Since the resulting authorization token can depend on the preexisting headers the `AuthorizationPolicy` must be executed right before the `TransportPolicy`. Also, since the token depends on the current timestamp (to mitigate reply attacks), `AuthorizationPolicy` must be run at **every** retry. 

# Cosmos context

In CosmosDB, the authorization token depends on the *resource type*. This information must be somehow passed through the pipeline because the `AuthorizationPolicy` needs it. 
The information of which resource to authenticate must be passed by the operation builder *prior* starting the pipeline. For example, the `create_database` operation, knowing that it has to work on `ResourceType::Databases`, must be able to pass `ResourceType::Databases` to the `AuthorizationPolicy`. The pipeline has a `Context` struct that must be instantiated by the SDK user (ie the code that calls `create_database`). The `Context` will then traverse the pipeline. As such is a good candidate for the resource type information. 

For this reason, this PR proposes an additional type, called `PipelineContext` that wraps the user-supplied `Context` and adds the module-dependent field. This is achieved by exposing a generic (constrained to be `Send + Sync`). 

An alternative approach (shown here: https://github.com/MindFlavor/azure-sdk-for-rust/tree/cosmos/auth_policy%2Fdev) is to use Rust's runtime polymorphism in the form of `std::any::Any`. Basically `Context` will have a `Box<dyn Any>` field that can accept anything. It would be responsibility of the operation builder (ie the `create_database` function) to store the relevant information in the `Box<dyn Any>` instance. The `AuthorizationPolicy` will then `downcast_ref()` the `Any` instance and get the expected type. 
While this approach is easy to do, it lacks any kind of type safety and relies upon conventions (which is very unidiomatic in Rust). Also, it's runtime enforced only (another thing unidiomatic in Rust). 

The approach proposed in this PR, instead, is type safe and clearly self-explanatory.

## Notable changes

* Created the `PipelineContext<BAG>` discussed above (composition of `Context` and custom field of type `BAG`).
* Changed `Policy` trait to be generic on the `BAG` above. This allows policies that accept generic `PipelineContext<BAG>`. 
* Changed the pipeline to be generic on the `BAG` above. 
* Fixed the preexisting policies to accept the `BAG` generic (called `R` for brevity, this is something we might want to correct to avoid confusion).
* Implemented `Policy<CosmosContext>` for `AuthorizationPolicy`.
* Changed pipeline code in CosmosDB to support the `AuthorizationPolicy` (and to send the proper `PipelineContext<CosmosContext>` down the pipeline.
* Removed useless `cosmos_client::prepare_request2`.
* Created a new function in `cosmos_client` called `prepare_request_pipeline`. This function will **no longer perform authentication** since it will delegated to the pipeline policy `AuthorizationPolicy`.
* Moved all the authorization code from `CosmosClient` to `AuthorizationPolicy` (preexisting non-pipeline code has been rerouted pending removal).
* Removed `uninmplemented` from `add_as_header2` that caused E2E tests to fail (fix #324).
* Implemented `add_as_header2` for all the required types.

### CODE REVIEW

File: sdk/cosmos/src/authorization_policy.rs [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+use url::form_urlencoded;
+
+const TIME_FORMAT: &str = "%a, %d %h %Y %T GMT";
+const AZURE_VERSION: &str = "2018-12-31";
+const VERSION: &str = "1.0";
+
+// We can implement Debug without leaking secrets because `AuthorizationToken`
+// already masks the secure bits on its own.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct AuthorizationPolicy {
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:19]: We should add rustdocs here and in the rest of this module.

File: sdk/cosmos/src/authorization_policy.rs [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            .headers_mut()
+            .append(HEADER_DATE, HeaderValue::from_str(&time)?);
+        request
+            .headers_mut()
+            .append(HEADER_VERSION, HeaderValue::from_static(AZURE_VERSION));
+        request
+            .headers_mut()
+            .append(AUTHORIZATION, HeaderValue::from_str(&auth)?);
+
+        trace!("\n\nrequest =={:?}", request);
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:91]: Can we remove this?

File: sdk/cosmos/src/authorization_policy.rs [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        "users",
+        "permissions",
+        "attachments",
+        "pkranges",
+        "udfs",
+        "triggers",
+    ];
+
+    // store the element only if it does not end with dbs, colls or docs
+    let p = u;
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:116]: The use of the variable names `p` and `u` makes the code hard to follow.

File: sdk/cosmos/src/authorization_policy.rs [hunk 4]
```diff
@@ -0,0 +1,320 @@
...
+    let len = p.len();
+    for str_to_match in ENDING_STRINGS {
+        let end_len = str_to_match.len();
+
+        if end_len <= len {
+            let end_offset = len - end_len;
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:122]: Underflow is possible here which we might want to protect against.

File: sdk/cosmos/src/authorization_policy.rs [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+pub(crate) fn generate_authorization(
+    auth_token: &AuthorizationToken,
+    http_method: &http::Method,
+    resource_type: &ResourceType,
+    resource_link: &str,
+    time: &str,
+) -> String {
+    let string_to_sign = string_to_sign(http_method, resource_type, resource_link, time);
+    debug!(
+        "generate_authorization::string_to_sign == {:?}",
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:149]: I'm not so sure how helpful this is. If you think it's useful, perhaps using `trace!` would be better?

File: sdk/cosmos/src/authorization_policy.rs [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+            AuthorizationToken::Resource(_) => "resource",
+        },
+        VERSION,
+        match auth_token {
+            AuthorizationToken::Primary(key) =>
+                Cow::Owned(encode_str_to_sign(&string_to_sign, key)),
+            AuthorizationToken::Resource(key) => Cow::Borrowed(key),
+        },
+    );
+    debug!(
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:166]: Same as above

File: sdk/cosmos/src/authorization_policy.rs [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        "generate_authorization::string_to_sign == {:?}",
+        string_to_sign
+    );
+
+    let str_unencoded = format!(
+        "type={}&ver={}&sig={}",
+        match auth_token {
+            AuthorizationToken::Primary(_) => "master",
+            AuthorizationToken::Resource(_) => "resource",
+        },
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:158]: Can we move this match expressions to bindings outside the `format!` call?

File: sdk/cosmos/src/authorization_policy.rs [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+            ResourceType::PartitionKeyRanges => "pkranges",
+            ResourceType::UserDefinedFunctions => "udfs",
+            ResourceType::Triggers => "triggers",
+        },
+        resource_link,
+        time.to_lowercase()
+    )
+}
+
+fn encode_str_to_sign(str_to_sign: &str, key: &[u8]) -> String {
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:220]: Even though all these functions are private or will become private, I think they could all use more documentation. How auth works in Azure isn't the most straightforward thing.

File: sdk/cosmos/src/authorization_policy.rs [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    p
+}
+
+// TODO: make it private after pipeline migration
+pub(crate) fn generate_authorization(
+    auth_token: &AuthorizationToken,
+    http_method: &http::Method,
+    resource_type: &ResourceType,
+    resource_link: &str,
+    time: &str,
```
Comment (rylev) [inline sdk/cosmos/src/authorization_policy.rs:145]: Should this take a `chrono` type so that the caller doesn't need to know the exact format that needs to be passed?

File: sdk/cosmos/src/clients/cosmos_client.rs [hunk 10]
```diff
@@ -54,21 +48,37 @@ impl CosmosOptions {
...
 /// Create a Pipeline from CosmosOptions
-fn new_pipeline_from_options(options: CosmosOptions) -> Pipeline {
+fn new_pipeline_from_options(
+    options: CosmosOptions,
+    authorization_token: AuthorizationToken,
+) -> Pipeline<CosmosContext> {
+    let auth_policy: Arc<dyn azure_core::Policy<CosmosContext>> =
+        Arc::new(crate::AuthorizationPolicy::new(authorization_token));
+
+    let mut per_retry_policies = Vec::with_capacity(1);
```
Comment (rylev) [inline sdk/cosmos/src/clients/cosmos_client.rs:58]: This call to `with_capacity` doesn't really help. The call to push would reserve space on the heap anyway, and `with_capacity` doesn't reserve exact capacity anyway.

File: sdk/cosmos/src/clients/cosmos_client.rs [hunk 11]
```diff
@@ -177,28 +188,41 @@ impl CosmosClient {
...
+                &resource_type,
                 resource_link,
                 &time,
             )
         };
         self.prepare_request_with_signature(uri_path, http_method, &time, &auth)
     }
 
-    /// Prepares' an `azure_core::Request`.
+    /// Prepares' an `azure_core::Request`. This function will
```
Comment (rylev) [inline sdk/cosmos/src/clients/cosmos_client.rs:206]: nit: not sure if we need to make the comment lines so short. Means we quickly have lots of lines of comments.

File: sdk/storage/src/data_lake/properties.rs [hunk 12]
```diff
@@ -47,6 +47,27 @@ impl<'a, 'b> AddAsHeader for Properties<'a, 'b> {
...
+    fn add_as_header2(
+        &self,
+        request: &mut azure_core::Request,
+    ) -> Result<(), http::header::InvalidHeaderValue> {
+        // the header is a comma separated list of key=base64(value) see
+        // [https://docs.microsoft.com/rest/api/storageservices/datalakestoragegen2/filesystem/create#request-headers](https://docs.microsoft.com/rest/api/storageservices/datalakestoragegen2/filesystem/create#request-headers)
+        let mut s = String::new();
+        self.0.iter().for_each(|(k, v)| {
+            s.push_str(&format!("{}={},", k.as_ref(), base64::encode(v.as_ref())));
+        });
```
Comment (rylev) [inline sdk/storage/src/data_lake/properties.rs:60]: I believe if you do the following you don't need to remove the last byte below. 
```suggestion
        let s = self.0.iter().map(|(k, v)| {
            format!("{}={}", k.as_ref(), base64::encode(v.as_ref()))
        }).collect::<Vec<_>>().join(",");
```

File: sdk/core/src/pipeline_context.rs [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+/// responsibility to specify the BAG generics: it can be anything as
+/// long as it's Send and Sync. For example, Cosmos uses the
+/// PipelineContext to pass the ResourceType down to its
+/// AuthorizationPolicy.
+pub struct PipelineContext<BAG>
+where
+    BAG: Send + Sync,
+{
+    inner_context: Context,
+    bag: BAG,
```
Comment (rylev) [inline sdk/core/src/pipeline_context.rs:17]: Not sure I love the name `bag`. It's really the service specific context.

File: sdk/core/src/pipeline.rs [hunk 14]
```diff
@@ -24,12 +24,18 @@ use std::sync::Arc;
 /// cannot be enforced by code). All policies except Transport policy can assume there is another following policy (so
 /// self.pipeline[0] is always valid).
 #[derive(Debug, Clone)]
-pub struct Pipeline {
+pub struct Pipeline<R>
```
Comment (rylev) [inline sdk/core/src/pipeline.rs:27]: nit: maybe the generic `C` would be a better letter since it's the context.

Also, can we document what the generic is?

**Review by rylev — CHANGES_REQUESTED** [review]
I was skeptical of the proposal, but it's not too bad. I definitely think it's better than having an `Any` that service specific policies can downcast.  I just had a bunch of small notes.

**Comment by MindFlavor** [comment]
Thank you @rylev for taking the time to review this! I'll try to address your suggestions and push another commit.

File: sdk/core/src/headers/mod.rs [hunk 15]
```diff
@@ -36,9 +36,10 @@ impl TryFrom<&HeaderMap> for CommonStorageResponseHeaders {
 
 pub trait AddAsHeader {
     fn add_as_header(&self, builder: Builder) -> Builder;
-    fn add_as_header2(&self, _request: &mut crate::Request) {
-        unimplemented!()
-    }
+    fn add_as_header2(
```
[the remark this answers is not in the document]
Comment (MindFlavor) [reply sdk/core/src/headers/mod.rs:39]: Yes that was my guess as well. 
I'll add some documentation. While at it I will change the expected error code from `http::header::InvalidHeaderValue` to a tagged union supporting `http::header::InvalidHeaderName` as well.

File: sdk/cosmos/src/authorization_policy.rs (hunk 3 above)
Comment (MindFlavor) [reply sdk/cosmos/src/authorization_policy.rs:116]: Yes! I've just moved that code but it can definitely benefit from a rewrite. 
I'll get to it. 👍

File: sdk/cosmos/src/authorization_policy.rs (hunk 9 above)
Comment (MindFlavor) [reply sdk/cosmos/src/authorization_policy.rs:145]: Good idea. I've hidden the whole chrono datetime - and its representation - in a opaque struct called `TimeNonce`.

File: sdk/cosmos/src/clients/cosmos_client.rs (hunk 11 above)
Comment (MindFlavor) [reply sdk/cosmos/src/clients/cosmos_client.rs:206]: I have a vim plugin that inserts a cr to keep comments < 80 chars wide (yes, I'm old 🤷‍♂️). I agree it doesn't really matter since Rust's lines can easily go over that limit. What should we use as a limit?

File: sdk/storage/src/data_lake/properties.rs (hunk 12 above)
Comment (MindFlavor) [reply sdk/storage/src/data_lake/properties.rs:60]: I honestly blindly copied the preexisting code but your solution is brilliant! Love it ❤️!

File: sdk/core/src/pipeline_context.rs (hunk 13 above)
Comment (MindFlavor) [reply sdk/core/src/pipeline_context.rs:17]: It's a leftover of the "old", property bag based approach. You are right, it's not fitting here. I'll use C everywhere.

File: sdk/cosmos/src/authorization_policy.rs (hunk 4 above)
Comment (MindFlavor) [reply sdk/cosmos/src/authorization_policy.rs:122]: I've completely rewritten the function in a more functional manner. It should be safe now.

**Review by rylev — APPROVED** [review]
@MindFlavor looks good. Shall we merge?

**Comment by MindFlavor** [comment]
> @MindFlavor looks good. Shall we merge?

Yes, let's keep the ball rolling! 👍
